src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
The IControllerPropertyActivator, like the IControllerActivator, is an abstraction which allows for applications to provide container specific versions of a behavior.  This is fine for MVC, as they do not need a common pattern, but we do.  That is the fundamental difference here.  We are introducing an extensibility point that requires, from the container, grain activator, or whatever other abstractions we introduce, that the pattern be respected.  By using property injection rather than constructor injection, we're just moving the problem to properties rather than constructors, and offering a inferior experience to our users in the process.

The idea that this is how MVC solved this problem indicates a misunderstanding of the problem.  MVC does not have this problem.  They have no need for a common pattern for either constructor injection or property injection.  They need only a means to allow applications to provide their own, container specific, versions of these features, which they do through these abstractions.

Whether we introduce facets via property injection or constructor injection, we still violate what MVC does.  In this, however, I do not care.  Using established patterns is only valuable if they solve the same problem, which these do not.  Given a choice between providing clean, intuitive, and maintainable capabilities to our users or arbitrarily confining our solution to existing common patterns, I will choose a better user experience every time.